presto-spi/src/main/java/com/facebook/presto/spi/function/ComplexTypeFunctionDescriptor.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.common.Subfield.allSubscripts;
+import static com.facebook.presto.common.Utils.checkArgument;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Contains properties that describe how the function operates on Map or Array inputs.
+ */
+public class ComplexTypeFunctionDescriptor
+{
+    public static final List<String> mapAndArray = unmodifiableList(Arrays.asList("map", "array"));
+    public static final ComplexTypeFunctionDescriptor DEFAULT = new ComplexTypeFunctionDescriptor(
+            true,
+            emptyList(),
+            Optional.of(emptySet()),
+            Optional.of(ComplexTypeFunctionDescriptor::allSubfieldsRequired));
+
+    /**
+     * Indicates whether the function accessing subfields.
+     */
+    private final boolean isAccessingInputValues;
+
+    /**
+     * Set of indices of the function arguments containing map or array arguments. Those arguments are important because all accessed subfields collected so far relate only to
+     * those map or array arguments and will be passed only to those arguments during the expression analysis phase.
+     * If <code>argumentIndicesContainingMapOrArray</code> is <code>Optional.empty()</code>, it indicates that all function arguments are of the map or array types.
+     * For example, <code>CONCAT</code> function accepts the variadic argument containing only array arguments.
+     * If <code>argumentIndicesContainingMapOrArray</code> is an empty set, it indicates that all function arguments are of the map or array types and thus all the collected
```

Comment:
Based on comments, it looks like behaviour is same if its Optional.empty() vs Optional.of(empty set). Is that the case?